blog/2024-04-16-custom-components.md

```diff
@@ -0,0 +1,110 @@
+---
+author: Nikhil Rao
+date: 2024-04-16
+title: Custom Components
+description: Announcing our third party custom component ecosystem.
+image: /blog/python.png
+---
+
+```python exec
+from pcweb.pages.docs import wrapping_react, custom_components, styling, events, tutorial
+from pcweb.pages.docs.custom_components import custom_components as cc
+```
+
+We are excited to announce the launch of our [custom component ecosystem]({cc.path})! 
+ 
+Custom components allow developers to build their own components and share them with the Reflex community. This includes wrapping React components, as well as creating higher level Reflex components.
+
+### New Features:
+
+* A `reflex components` subcommand that allows you to create, test, and publish custom components.
+* Install third party components from PyPI with `pip`.
+* A new `rx.ComponentState` class that allows you to encapsulate state and UI for a component in a single class.
+
+## What are Custom Components?
+
+Reflex has many built-in components for common use cases, but sometimes you may want to create your own components. Custom components allow you to encapsulate complex logic and styling into a single reusable component.
+
+For example, in our [tutorial]({tutorial.final_app.path}) we share the code for creating a chat component in Reflex. In order to build a chat app, you would have to copy and paste the code for the chat component into your own app. 
+
+With custom components, we now have a [reflex-chat](https://github.com/picklelo/reflex-chat/) package that you can install with `pip` and use in your app.
+
+```bash
+pip install reflex-chat
+```
+
+You can then import the chat component into your app and use it like any other Reflex component. All the styling and logic is encapsulated, you only need to pass in the actual LLM logic.
+
+```python exec
+import reflex as rx
+from openai import OpenAI
+from reflex_chat import chat
+
+client = OpenAI()
+
+# Only define your logic, the chat component handles the rest.
+async def process(chat):
```

Comment:
shadowing the imported name `chat` makes the code example a little harder to grok.

consider calling this arg something like `chat_state` so distinguish it from the component used in the UI code.

Reply:
Good call, changed this